compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/AggregatedOpInterfaceImpl.cpp

```diff
@@ -246,18 +326,69 @@ OnlineAttentionOp::decomposeOperation(OpBuilder &b) {
   AffineMap sumMap = getSumMap();
   Value normSum = scaleValueInPlace(b, loc, sumMap, normMap, oldSum, norm);
 
-  // newSum = normSum + rowMax(P)
+  // newSum = normSum + rowSum(P)
   Value newSum = reduce<arith::AddFOp>(b, loc, pMap, sumMap, p, normSum);
 
   // newAcc = norm * oldAcc
   AffineMap accMap = getOutputMap();
+
+  // ---- Scale and truncate LHS to match RHS ----
+  Value pScale;
+  auto pETy = getElementTypeOrSelf(p.getType());
+  if (pETy != vETy && isa<FloatType>(vETy)) {
+    if (vETy.getIntOrFloatBitWidth() <= 8) {
+      SmallVector<OpFoldResult> mSizes;
+      for (AffineExpr dimExpr : maxMap.getResults()) {
+        int dim = cast<AffineDimExpr>(dimExpr).getPosition();
+        mSizes.push_back(sizes[dim]);
+      }
+
+      // We rescale `p` to use the full range and pass back the `pScale`.
+      Value maxEmpty = b.create<tensor::EmptyOp>(loc, mSizes, reduceType);
+      Value absMax =
+          reduce<arith::MaximumFOp>(b, loc, pMap, maxMap, p, maxEmpty);
+
+      auto fpTy = cast<FloatType>(vETy);
+      double largestDbl =
+          APFloat::getLargest(fpTy.getFloatSemantics(), /*Negative=*/false)
+              .convertToDouble();
+
+      AffineMap scaleMap = AffineMap::get(/*dimCount=*/maxMap.getNumInputs(),
+                                          /*symbolCount=*/0, getContext());
+
+      // Compute the post matmul scale required:
+      Value largestInv = b.create<arith::ConstantOp>(
```

Comment:
Can we get better high-level comment on why we are doing this?

Reply:
Added more comments.